osu.Game/Rulesets/Difficulty/Utils/DropoutStack.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    /// <summary>
+    /// An indexed stack where items are dropped from the bottom rather than popped from the top.
+    /// Indexing starts at the top of the <see cref="DropoutStack{T}"/>.
+    /// </summary>
```

Comment:
Not overly sure about the chosen semantics. What this is is essentially a queue, but with the "indexing reversed" (tail of queue has smallest index? I think?). Might as well turn it into a deque at this point if we're going for flexibility? Or just plain list? Not sure.

Definitely can say that performance-wise it is definitely worse, but not sure by how much (see the `O(n)` implementation of `Drop()` - popping off the `LimitedStack` was `O(1)`).

Reply:
Good points.

I've addressed this with 4eb125374f7c0589cf5a6312e02e68785a265ab9 by replacing the `DropoutStack` with a `ReverseQueue`.
Is has essentially the same functionality, but with a more semantically appropriate name and lower level implementation allowing for O(1) enqueue*, dequeue, and access operations.
*The only exception is enqueue being O(n) when it hits its internal capacity and resizes to double the buffer size.
This O(n) case will only be a possibility for skills with truly unknown size requirements however because of the configurable initial capacity added in 3cf8635c61f3a84056e833975e34f6f277671f5e.